### DISCUSSION TITLE
Modifying to work with I2C?

### DISCUSSION THREAD

**Discussion by zandernc** [Spirik/GEM#72] [Q&A] [answered]
After spending a couple days looking for a menu system I could use in a little project - GEM seems the best documented and easiest to understand (I'm a rank novice with coding). Your organization and Wiki are the best I've seen! My stumbling block is I am trying to use a display that communicates over I2C. Is that a huge modification? 

Specifically, I'm working on the Adafruit Feather M0 with a 128 x 64 old "featherwing"
https://learn.adafruit.com/adafruit-feather-m0-basic-proto
https://learn.adafruit.com/adafruit-128x64-oled-featherwing

**Comment by Spirik** [comment]
👋 Should be no problem to operate your display, as soon as it is supported by one of the graphical libraries that GEM can work with (namely [AltSerialGraphicLCD](http://www.jasspa.com/serialGLCD.html), [U8g2](https://github.com/olikraus/U8g2_Arduino) and [Adafruit GFX](https://learn.adafruit.com/adafruit-gfx-graphics-library)). In your case it is (quite expectedly) [supported](https://learn.adafruit.com/adafruit-128x64-oled-featherwing/arduino-code) by Adafruit GFX library.

I'd recommend to finish Adafruit [tutorial](https://learn.adafruit.com/adafruit-128x64-oled-featherwing/arduino-code) with its basic sketch (just to make sure the display is working correctly) and probably run some other examples that come with Adafruit GFX library.

Then supply object constructed with Adafruit GFX library (e.g. `Adafruit_SH1107 display = Adafruit_SH1107(64, 128, &Wire);`) to GEM constructor (e.g. `GEM_adafruit_gfx menu(display, GEM_POINTER_ROW, GEM_ITEMS_COUNT_AUTO);` ).

Next you would probably need to set correct orientation of the display (according to the Adafruit example `display.setRotation(1);` should do the trick) and [adjust](https://github.com/Spirik/GEM/wiki/Menu-appearance) different sizes and offsets to your liking.

And probably run some basic Adafruit GFX specific [examples](https://github.com/Spirik/GEM/tree/master/examples/AdafruitGFX) that come with GEM.

Hope that helps!

**Comment by zandernc** [comment]
Wow - thank you so much for the response! I do have the display working with other sketches... but I have been trying to modify your simple Example-01_Basic.ino sketch to work with my display and something isn't working. I think I'm confused about what do do with the lines related to communicating with the display: 
```
#define TFT_CS    A2
#define TFT_RST   -1 // Set to -1 and connect to Arduino RESET pin
#define TFT_DC    A3
```
and ```tft.initR(INITR_BLACKTAB);``` which I replaced with ```tft.begin(0x3C, true);``` but I'm not getting anything to show on the display.

I thought it might be something to do with color settings? I'm not sure. I'd be happy to share the full sketch if that's helpful...

**Reply by Spirik** [reply]
Probably something along those lines will work: https://gist.github.com/Spirik/5e3ce403a49f42c1c066989b0342b657 .

I copied a bunch of initialization options from Adafruit [example](https://learn.adafruit.com/adafruit-128x64-oled-featherwing/arduino-code), not sure if all of those are necessary. Also I mapped three of the available buttons to Ok (BUTTON_A), Cancel (BUTTON_B) and Down (BUTTON_C). Should be enough for basic navigation (although without an ability to properly edit variables).

At least it compiled without errors=)

**Reply by zandernc** [reply]
Thank you so much! You are amazing to do this for me. Sadly, I tried that sketch and the display is still blank. The Adafruit splash screen comes up on reset, but then goes blank. I can see the menu is running, though, as blindly navigating the menu I can get the serial monitor message "Printing is disabled, sorry:(" Any other thoughts?

**Reply by Spirik** [reply]
What will happen if you add another call to `display.display();` after `menu.drawMenu();` (at the very end of `setup()`)?

UPD: it seems that these series of displays utilize an internal buffer, which stores every drawing operation, results of which won't be visible on an actual screen until [`display.display()`](https://adafruit.github.io/Adafruit_SH110x/html/class_adafruit___s_h110_x.html#af8b0e8c8c4cbad863aa23859ea70622a) command is called:
> Drawing operations are not visible until this function is called. Call after each graphics command, or after a whole set of graphics commands, as best needed by one's own application.

GEM internally doesn't call it, so it must be done in sketch. At the end of `setup()` and presumably somewhere in the `loop()` (probably after `menu.registerKeyPress(myKeyDetector.trigger)`).

**Reply by zandernc** [reply]
Well I tried that - I added it to the end of the last bit of the ```setup() section```.
```
  // Menu init, setup and draw
  menu.init();
  setupMenu();
  menu.drawMenu();
  display.display();
}
```
and at the end of ```void loop()```

```
void loop() {
  // If menu is ready to accept button press...
  if (menu.readyForKey()) {
    // ...detect key press using KeyDetector library
    myKeyDetector.detect();
    // Pass pressed button to menu
    // (pressed button ID is stored in trigger property of KeyDetector object)
    menu.registerKeyPress(myKeyDetector.trigger);
    display.display();
  }
}
```
but still nothing. I tried adding it after that last to next bracket too - no difference. I thought that was going to do it, too!

**Reply by Spirik** [reply, marked as answer]
Let's try specifying colors for the display explicitly. Before `menu.init()` add:
```cpp
menu.setForegroundColor(SH110X_WHITE);
menu.setBackgroundColor(SH110X_BLACK);
```
And keep `display.display()` in your sketch (especially somewhere in `loop()`).

**Reply by zandernc** [reply]
YESSS!!! That did it. (I tried those color lines earlier but with out the extra ```display.display()```'s .  Thank you so much for your patience and help with this! So yes - the combination of the color definitions and the draw display calls are what did it. (clarity for posterity!)

**Reply by Spirik** [reply]
That's great news! Glad I was able to help. I will include all necessary updates in sketch gist and link to it from wiki.

But before that, one more thing to clarify though. I assume [GEM splash screen](https://github.com/Spirik/GEM/wiki/Splash-screen#gem-splash) won't come up on display no mater what: you get Adafruit splash and then (some couple of seconds later) menu is drawn, is that correct?

**Reply by zandernc** [reply]
That is correct - it would be cool to have the GEM splash screen come up. 

One other thing just FYI: I was not able to get the program to compile with this board (the Adafruit Feather M0) without commenting out the includes for AltSerial and U8g2 in config.h. (which I, of course, forgot about, so when GEM updated, nothing was working! Doh!)

**Reply by Spirik** [reply]
The thing is, internally GEM draws its splash (and version number alongside it) and clears screen afterwards in `menu.init()` routine, so in order for it to be actually visible on buffer-equiped display (such as this one) `display.display()` should be called right there as well. Which is kind of an update to library that I am not sure is safe to make for obvious compatibility reasons: not all of the screens may have support for `display.display()` method, and I don't know if there is a way to reliable test for that.

So I'd rather suggest you to disable GEM splash and version number altogether by adding the following line (before `menu.init()`):
```cpp
menu.setSplashDelay(0);
```
On the brighter side, that will speed up loading for a second. And even more so, if you disable Adafruit splash as well, by commenting out the very first call to `display.display()` in `setup()`:
```cpp
// Show image buffer on the display hardware.
// Since the buffer is intialized with an Adafruit splashscreen
// internally, this will display the splashscreen.
// display.display();
// delay(1000);
```

I've updated [example gist](https://gist.github.com/Spirik/5e3ce403a49f42c1c066989b0342b657) with these changes.

And yeah, commenting out unnecessary libraries in `config.h` is pain in a butt. It's much easier in some dev environments (such as PlatformIO) that provide a way to specify compiler [build flags](https://github.com/Spirik/GEM#defining-build-flags) on per sketch basis (without the need to mess with library source code). Hopefully, something similarly convenient will be added in Arduino IDE 2 some time in the future.

**Comment by zandernc** [comment]
Tested - this works! Thanks again!
